packages/slice-machine/components/ScreenshotChangesModal/index.tsx

```diff
@@ -0,0 +1,205 @@
+import { useState } from "react";
+
+const VariationIcon: React.FC<{ isValid?: boolean }> = ({ isValid }) => (
+  <Flex
+    sx={{
+      p: 2,
+      border: "1px solid #E4E2E4",
+      borderRadius: "6px",
+      bg: "#FFF",
+    }}
+  >
+    {isValid ? (
+      <AiOutlinePicture style={{ fontSize: "16px" }} />
+    ) : (
+      <RiErrorWarningLine style={{ fontSize: "16px", color: "#ED811C" }} />
+    )}
+  </Flex>
+);
+
+const VariationsList = ({
+  slice,
+  onSelectVariation,
+  selectedSliceVariation,
+}: {
+  slice: ComponentUI;
+  isSelected?: boolean;
+  selectedSliceVariation: [string, string];
+  onSelectVariation: (sliceID: string, variationID: string) => void;
+}) => (
+  <>
+    <Heading
+      as="h4"
+      sx={{
+        whiteSpace: "nowrap",
+        overflow: "hidden",
+        textOverflow: "ellipsis",
+      }}
+    >
+      {slice.model.name}
+    </Heading>
+    <Box>
+      <Box
+        as="ul"
+        sx={{
+          pl: 0,
+          my: 3,
+        }}
+      >
+        {slice.model.variations.map((variation) => {
+          const isSelectedVariation =
+            selectedSliceVariation[0] === slice.model.id &&
+            selectedSliceVariation[1] === variation.id;
+          return (
+            <Box
+              as="li"
+              sx={{
+                listStyle: "none",
+                my: 2,
+                p: 2,
+                cursor: "pointer",
+                borderRadius: "4px",
+                ...(isSelectedVariation && {
+                  bg: "#EDECEE",
+                  fontWeight: "500",
+                }),
+                ":hover": {
+                  bg: "#EDECEE",
+                },
+              }}
+              onClick={() => onSelectVariation(slice.model.id, variation.id)}
+            >
+              <Flex
+                sx={{
+                  alignItems: "center",
+                }}
+              >
+                <VariationIcon isValid={!!slice.screenshots[variation.id]} />
+                <Text sx={{ ml: 2 }}>{variation.name}</Text>
+              </Flex>
+            </Box>
+          );
+        })}
+      </Box>
+    </Box>
+  </>
+);
+
+const ScreenshotChangesModal = ({
+  slices,
+  defaultSelectedSliceVariation = [],
+  onClose,
+}: {
+  slices: ComponentUI[];
+  defaultSelectedSliceVariation?: [string, string] | [];
+  onClose: () => void;
+}) => {
+  const { isOpen } = useSelector((store: SliceMachineStoreType) => ({
+    isOpen: isModalOpen(store, ModalKeysEnum.SCREENSHOTS),
+  }));
+  const [defaultSelectedSliceID, defaultSelectedSliceVariationID] =
+    defaultSelectedSliceVariation;
+
+  const [selectedSliceVariation, setSelectedSliceVariation] = useState<
+    [string, string]
+  >(
+    slices.some(
+      (s) =>
+        s.model.id === defaultSelectedSliceID &&
+        s.model.variations.some((v) => v.id === defaultSelectedSliceVariationID)
+    )
+      ? (defaultSelectedSliceVariation as [string, string])
```

Comment:
weird, you know type-gaurds can help get rid o the need to cast you'll can find a few examples by looking for functions using the `is` keyword

Reply:
Can we maybe provide some different types here to clarify a bit?
```
slices: ComponentUI[]
defaultSelectedSliceVariation?: [string, string] | []
```
It could maybe become something like:
```
slices: NonEmptyArray<ComponentUI>
defaultSelectedSliceVariation?: [string, string] | undefined
```
It would ensure that you first check that `defaultSelectedSliceVariation` exists before destructuring and so no need for casting later on.
Also if undefined, you fallback with an indexed access on slices that could be an empty array, If you can be absolutely sure it won’t be empty, you can encode it in the type with my suggestion above, WDYT @hypervillain ?
I think the cast here is not the issue, it’s more about how far the types are from the actual truth about what the app will provide.